unit_test/test_spdm_callback/spdm_cert_verify_callback.c

```diff
@@ -0,0 +1,253 @@
+    result = libspdm_asn1_get_tag(&ptr, end, &obj_len,
+                                  LIBSPDM_CRYPTO_ASN1_CONTEXT_SPECIFIC + 8);
+    if (result) {
+        ptr += obj_len;
+    }
+    /*type*/
+    result = libspdm_asn1_get_tag(&ptr, end, &obj_len,
+                                  LIBSPDM_CRYPTO_ASN1_CONTEXT_SPECIFIC + 9);
+    if (result) {
+        if ((obj_len != sizeof(m_libspdm_dice_tcbinfo_type)) ||
+            (!libspdm_consttime_is_mem_equal(ptr, m_libspdm_dice_tcbinfo_type, obj_len))) {
+            return false;
+        }
+        ptr += obj_len;
+    }
+    /*flagMask*/
+    result = libspdm_asn1_get_tag(&ptr, end, &obj_len,
+                                  LIBSPDM_CRYPTO_ASN1_CONTEXT_SPECIFIC + 10);
+    if (result) {
+        ptr += obj_len;
+    }
+
+    if (ptr == end) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/*callback function for verifying cert_chain DiceTcbInfo extension*/
+bool libspdm_verify_spdm_cert_chain_with_dice(void *spdm_context, uint8_t slot_id,
+                                              size_t cert_chain_size, const void *cert_chain,
+                                              const void **trust_anchor,
+                                              size_t *trust_anchor_size)
+{
+    bool result;
+    libspdm_context_t *context;
+    const uint8_t *cert_chain_data;
+    size_t cert_chain_data_size;
+    size_t hash_size;
+    uint8_t *ptr;
+    uint8_t *tem_ptr;
+    int32_t length;
+    size_t obj_len;
+    uint8_t *end;
+
+    /*verify peer cert chain integrity*/
+    result = libspdm_verify_peer_cert_chain_buffer_integrity(spdm_context, cert_chain,
+                                                             cert_chain_size);
+    if (!result) {
+        return false;
+    }
+
+    /*verify peer cert chain authority*/
+    result = libspdm_verify_peer_cert_chain_buffer_authority(spdm_context, cert_chain,
+                                                             cert_chain_size, trust_anchor,
+                                                             trust_anchor_size);
+    if (!result) {
+        return false;
+    }
+
+    context = spdm_context;
+    hash_size = libspdm_get_hash_size(context->connection_info.algorithm.base_hash_algo);
+
+    cert_chain_data = (const uint8_t *)cert_chain + sizeof(spdm_cert_chain_t) + hash_size;
+    cert_chain_data_size = cert_chain_size - sizeof(spdm_cert_chain_t) - hash_size;
+
+    length = (int32_t)cert_chain_data_size;
+    ptr = (uint8_t*)(size_t)cert_chain_data;
+    obj_len = 0;
+    end = ptr + length;
+
+    while (ptr < end) {
+        tem_ptr = ptr;
+        result = libspdm_asn1_get_tag(&ptr, end, &obj_len,
+                                      LIBSPDM_CRYPTO_ASN1_SEQUENCE |
+                                      LIBSPDM_CRYPTO_ASN1_CONSTRUCTED);
+        if (result) {
+            /*verify Dice TCB info*/
+            result = libspdm_verify_cert_dicetcbinfo(tem_ptr, obj_len + (ptr - tem_ptr));
```

Comment:
The "m_libspdm_must_have_dice_tcb_info" is a global that applies to all certs in the chain, but TcbInfo is usually present and required for some certs (e.g., Alias cert).

Reply:
Please note, this is reference implementation. We cannot handle all difference use cases.

The integrator should choose what to do in the final production, based on their own use case.